worker/src/services/IngestionService/index.ts

```diff
@@ -418,54 +419,56 @@ export class IngestionService {
   }): Promise<
     | Pick<
         ObservationRecordInsertType,
-        | "input_usage_units"
-        | "output_usage_units"
-        | "total_usage_units"
-        | "input_cost"
-        | "output_cost"
-        | "total_cost"
-        | "internal_model_id"
+        "usage_details" | "cost_details" | "total_cost" | "internal_model_id"
       >
     | {}
   > {
     const { projectId, observationRecord } = params;
-
     const internalModel = await findModel({
       event: {
         projectId,
         model: observationRecord.provided_model_name ?? undefined,
-        unit: observationRecord.unit ?? undefined,
       },
     });
 
-    const tokenCounts = this.getTokenCounts(observationRecord, internalModel);
-    const tokenCosts = IngestionService.calculateTokenCosts(
-      internalModel,
+    const final_usage_details = this.getUsageUnits(
+      observationRecord,
+      internalModel
+    );
+    const modelPrices = await this.getModelPrices(internalModel?.id);
+    const final_cost_details = IngestionService.calculateUsageCosts(
+      modelPrices,
       observationRecord,
-      tokenCounts
+      final_usage_details.usage_details ?? {}
     );
 
     return {
-      ...tokenCounts,
-      ...tokenCosts,
+      ...final_usage_details,
+      ...final_cost_details,
       internal_model_id: internalModel?.id,
-      unit: observationRecord.unit ?? internalModel?.unit,
     };
   }
 
-  private getTokenCounts(
+  private async getModelPrices(modelId?: string): Promise<Price[]> {
+    return modelId
+      ? ((await this.prisma.price.findMany({ where: { modelId } })) ?? [])
```

Comment:
do we also need a project_id check on the models table here? Either matching project_id or null

Reply:
Each price is already linked to a `model_id` which in turn is linked to a project or null. Would be duplicate here and some of the prisma queries do not work well with project_id null in a composite key